src/utils/file/file_utils.cpp

```diff
@@ -0,0 +1,247 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <errno.h>
+#include <file_utils.h>
+#include <filesystem>
+#include <algorithm>
+#include <fcntl.h>
+#include "common/nixl_log.h"
+#include <absl/strings/str_format.h>
+
+int nixlFileUtils::getModeFlags(nixlFileMode mode, bool use_direct) {
+    // Base flags for each mode
+    switch (mode) {
+        case nixlFileMode::CREATE:
+        case nixlFileMode::CREATE_DIRECT:
+            return ((mode == nixlFileMode::CREATE_DIRECT || use_direct) ? O_DIRECT : 0) |
+                   O_CREAT | O_WRONLY | O_TRUNC;
+
+        case nixlFileMode::READ_ONLY:
+        case nixlFileMode::READ_DIRECT:
+            return ((mode == nixlFileMode::READ_DIRECT || use_direct) ? O_DIRECT : 0) |
+                   O_RDONLY;
+
+        case nixlFileMode::WRITE_ONLY:
+        case nixlFileMode::WRITE_DIRECT:
+            return ((mode == nixlFileMode::WRITE_DIRECT || use_direct) ? O_DIRECT : 0) |
+                   O_WRONLY;
+
+        case nixlFileMode::READ_WRITE:
+        case nixlFileMode::READ_WRITE_DIRECT:
+            return ((mode == nixlFileMode::READ_WRITE_DIRECT || use_direct) ? O_DIRECT : 0) |
+                   O_RDWR | O_CREAT;
+
+        default:
+            return -1;
+    }
+}
+
+nixl_status_t nixlFileUtils::openFile(const std::string& path, nixlFileMode mode, int& fd, bool use_direct) {
+    int flags = getModeFlags(mode, use_direct);
+    if (flags == -1) {
+        NIXL_ERROR << absl::StrFormat("Invalid mode flags for path: %s", path);
+        return NIXL_ERR_INVALID_PARAM;
+    }
+
+    // For WRITE_ONLY, check if file exists and add O_CREAT if it doesn't
+    if (mode == nixlFileMode::WRITE_ONLY) {
+        nixl_status_t status = fileExists(path);
```

Comment:
With this check, the code is prone to the [Time-of-check to time-of-use](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issue. One process can get `true` on the check, followed by `rename` or `unlink` by another process, then the first process will fail creating the file.

I think it is better to be consistent with `READ_WRITE` mode, always provide `O_CREAT` for write, and remove this check.